## 05-7 외부 라이브러리
파이썬 설치 시 기본으로 설치되는 라이브러리를 파이썬 표준 라이브러리라고 한다. 이번에 소개하는 외부 라이브러리는 파이썬 표준 라이브러리가 아니므로 사용하려면 먼저 pip 도구를 이용하여 설치해야 한다.

> pip은 "핍" 이라 읽는다.

### pip
pip은 파이썬 모듈이나 패키지를 쉽게 설치할 수 있도록 도와주는 도구이다. pip으로 파이썬 프로그램을 설치하면 의존성 있는 모듈이나 패키지를 함께 설치해 주기 때문에 매우 편리하다. 예를 들어 B라는 파이썬 패키지를 설치하려면 A라는 패키지가 먼저 설치되어야 하는 규칙이 있다고 할 때 pip을 이용하면 B 패키지를 설치할 때 A 패키지도 자동으로 함께 설치된다.

pip 사용법에 대해서 간략하게 알아보자.

#### pip install
PyPI(Python Package Index)는 파이썬 소프트웨어가 모인 저장 공간이다. 현재 이곳에는 100,000건 이상의 파이썬 패키지가 등록되어 있으며 이곳에 등록된 파이썬 패키지는 누구나 내려받아 사용할 수 있다. 이곳에서 직접 내려받아 설치해도 되지만 pip을 이용하면 다음과 같이 간편하게 설치를 할 수 있다.

```bash
pip install SomePackage 

여기서 SomePackage는 내려받을 수 있는 특정 패키지를 뜻한다.

#### pip uninstall
설치한 패키지를 삭제하고 싶다면 다음의 명령어로 삭제할 수 있다.

```bash
pip uninstall SomePackage
```
#### 특정 버전으로 설치하기
다음과 같이 버전을 지정하여 설치할 수도 있다. 다음 명령어를 실행하면 1.0.4 버전의 SomePackage를 설치한다.
```bash
pip install SomePackage==1.0.4
다음처럼 버전을 생략하면 최신 버전을 설치한다.

pip install SomePackage 

#### 최신 버전으로 업그레이드하기
패키지를 최신 버전으로 업그레이드하려면 --upgrade 옵션과 함께 사용한다.

pip install --upgrade SomePackage 

설치된 패키지 확인하기
다음 명령은 pip을 이용하여 설치한 패키지 목록을 출력한다.
```bash
pip list
다음과 같이 설치된 패키지 목록을 출력할 것이다.

Package                  Version  
------------------------ --------  
amqp                       2.1.4  
anyjson                    0.3.3  
billiard                   3.3.0.23  
celery                     3.1.0  
defusedxml                 0.4.1  
diff-match-patch           20121119  
(... 생략 ...)  

### Faker
이번에는 pip을 사용하여 유용한 외부 라이브러리중 하나인 faker를 설치하고 사용해 보자. faker는 테스트용 가짜 데이터를 생성할 때 사용하는 라이브러리이다.

faker 라이브러리는 pip을 이용하여 설치해야 한다.

```bash
C:\> pip install Faker
```

#### Faker 사용해 보기
만약 다음과 같은 형식의 테스트 데이터 30건이 필요하다고 가정해 보자. 직접 데이터를 작성하지 말고 좀 더 편리한 방법으로 테스트 데이터를 만들려면 어떻게 해야 할까?
>[(이름1, 주소1), (이름2, 주소2), ..., (이름30, 주소30)]

테스트 데이터는 faker를 사용하면 아주 쉽게 만들 수 있다. 이름은 다음처럼 만들 수 있다.



In [15]:
from faker import Faker
fake = Faker()
fake.name()

'William Johnson'

한글 이름이 필요하다면 다음과 같이 한국을 의미하는 ko-KR을 전달하여 fake 객체를 생성하면 된다.



In [16]:
fake = Faker('ko_KR')
fake.name()

'최상철'

주소는 다음과 같이 만들 수 있다.

In [17]:
fake.address()

'전라남도 동해시 도산대2거리 (재현윤한마을)'

따라서 이름과 주소를 쌍으로 하는 30건의 테스트 데이터는 다음과 같이 만들 수 있다.

In [18]:
test_data = [(fake.name(), fake.email(), fake.password()) for i in range(10)]

In [19]:
test_data

[('우지혜', 'junhyeogi@example.org', 'p4XHjQEq0@'),
 ('최재현', 'ihyeonju@example.net', '96IMdlHc@1'),
 ('박영순', 'jeongsug93@example.org', '^8aOnKveGJ'),
 ('최종수', 'gimhyeonji@example.com', '^Bg1VJf%Wn'),
 ('이광수', 'lgim@example.net', '*c6PH#pZQT'),
 ('김지후', 'jihyeon23@example.com', '*(4M*d1s@r'),
 ('이재호', 'bgim@example.org', '_M5QpO(q_u'),
 ('문은경', 'jeongungbag@example.com', 'o4g*TzpF#I'),
 ('김정호', 'eunjimun@example.com', '@&0UHshmOn'),
 ('홍정남', 'subin56@example.org', '8(2LMdsEFb')]

In [20]:
test_data = [(fake.name(), fake.address()) for i in range(30)]

In [21]:
test_data

[('강영철', '울산광역시 성동구 언주729길 (성현유김동)'),
 ('김민석', '경상북도 시흥시 삼성길 (하윤김이면)'),
 ('류상훈', '전라북도 아산시 잠실거리 (성훈권김리)'),
 ('류혜진', '서울특별시 동대문구 석촌호수가 (준혁이성동)'),
 ('류도윤', '경상북도 예산군 백제고분로'),
 ('이경희', '전라북도 고양시 덕양구 가락거리 (영식김읍)'),
 ('권성수', '경상북도 화천군 서초대168로 (중수류동)'),
 ('진윤서', '충청북도 용인시 기흥구 테헤란길 (정남이면)'),
 ('권정훈', '제주특별자치도 음성군 봉은사52거리 (현준오리)'),
 ('한은지', '광주광역시 성북구 압구정8가'),
 ('김민수', '충청북도 의정부시 봉은사044가'),
 ('진하윤', '전라북도 안양시 동안구 가락거리'),
 ('김혜진', '대전광역시 종로구 반포대6길 (도현이마을)'),
 ('안영일', '부산광역시 금천구 선릉95가'),
 ('김서연', '대구광역시 동구 강남대1길 (민서이읍)'),
 ('류명숙', '경기도 공주시 봉은사거리'),
 ('장혜진', '인천광역시 송파구 가락3로 (하윤한장면)'),
 ('임지민', '경기도 제천시 서초중앙7로'),
 ('김준영', '전라남도 부천시 오정구 서초중앙가 (지아박김마을)'),
 ('박서현', '광주광역시 구로구 서초중앙68로 (현준최이마을)'),
 ('이영식', '대전광역시 동구 도산대76가 (상훈이강마을)'),
 ('오은주', '대구광역시 금천구 양재천8로'),
 ('이민서', '광주광역시 성북구 압구정가 (미숙이김읍)'),
 ('박영순', '경상남도 음성군 서초대가'),
 ('이영호', '대구광역시 금천구 학동39로'),
 ('이재호', '광주광역시 영등포구 개포28가'),
 ('한영수', '경상남도 수원시 영통구 서초중앙가 (준혁이면)'),
 ('김지훈', '서울특별시 서대문구 논현699가'),
 ('강현숙', '강원도 동두천시 반포대41거리'),
 ('김혜진', '인천광역시 용산구 서

#### Faker 활용
faker는 앞서 살펴본 name, address 이외에 다른 항목도 제공한다. 대표적인 몇 가지만 알아보자.

```
항목	설명
fake.name()	이름
fake.address()	주소
fake.postcode()	우편 번호
fake.country()	국가명
fake.company()	회사명
fake.job()	직업명
fake.phone_number()	휴대 전화 번호
fake.email()	이메일 주소
fake.user_name()	사용자명
fake.pyint(min_value=0, max_value=100)	0부터 100 사이의 임의의 숫자
fake.ipv4_private()	IP 주소
fake.text()	임의의 문장 (한글 임의의 문장은 fake.catch_phrase() 사용)
fake.color_name()	색상명
```

### sympy
sympy는 방정식 기호(symbol)를 사용하게 해주는 외부 라이브러리이다. 마찬가지로 pip을 이용하여 sympy를 설치하자.
```bash
C:\> pip install sympy


#### sympy 사용해 보기
시윤이는 가진 돈의 2/5로 학용품을 샀다고 한다. 이때 학용품을 사는 데 쓴 돈이 1,760원이라면 남은 돈은 어떻게 구하면 될까?

이 문제는 연습장과 연필만 있으면 쉽게 구할 수 있는 일차방정식 문제이다. 파이썬으로는 다음처럼 sympy를 사용하면 방정식을 쉽게 풀 수 있다. 먼저 다음과 같이 fractions 모듈과 sympy 모듈이 필요하다.

In [22]:
from fractions import Fraction
import sympy

시윤이가 가진 돈을 x라 하면 sympy 모듈을 사용하여 다음과 같이 표현할 수 있다.
```python
>>> x = sympy.symbols("x")

sympy.symbols()는 x처럼 방정식에 사용하는 미지수를 나타내는 기호를 생성할 때 사용한다.

여러개의 기호 사용하기
x, y 두 개의 미지수가 필요하다면 다음처럼 표현할 수 있다.

x, y = sympy.symbols('x y')

시윤이가 가진 돈의 2/5
가 1,760원, 즉 일차방정식 x * (2/5) = 1,760원 
이므로 이를 코드로 표현하면 다음과 같다.

In [23]:
x, y = sympy.symbols('x y')
f = sympy.Eq(x*Fraction('2/5'), 1760)
f

Eq(2*x/5, 1760)

sympy.Eq(a, b)는 a와 b가 같다는 방정식이다. 여기서 사용한 Fraction은 유리수를 표현할 때 사용하는 파이썬 클래스로, 2/5를 정확하게 계산하고자 사용했다.


#### fractions.Fraction
파이썬에서 유리수 연산을 정확하게 하려면 fractions.Fraction을 사용해야 한다.

```python
>>> from fractions import Fraction
유리수는 다음처럼 Fraction(분자, 분모) 형태로 만들 수 있다.

In [24]:
a = Fraction(1, 5)
a

Fraction(1, 5)

In [25]:
a = Fraction('1/5')
a

Fraction(1, 5)

f라는 방정식을 세웠으므로 sympy.solve(f)로 x에 해당하는 값을 구할 수 있다.

In [26]:
result = sympy.solve(f)
result

[4400]

sympy.Eq(a, b)는 a와 b가 같다는 방정식이다. 여기서 사용한 Fraction은 유리수를 표현할 때 사용하는 파이썬 클래스로, 2/5를 정확하게 계산하고자 사용했다.

In [27]:
remains = result[0] - 1760
remains

2640

```python
from fractions import Fraction
import sympy

# 가지고 있던 돈을 x라고 하자.
x = sympy.symbols("x")

# 가지고 있던 돈의 2/5가 1760원이므로 방정식은 x * (2/5) = 1760 이다.
f = sympy.Eq(x*Fraction('2/5'), 1760)

# 방정식을 만족하는 값(result)을 구한다.
result = sympy.solve(f)  # 결괏값은 리스트

# 남은 돈은 다음과 같이 가지고 있던 돈에서 1760원을 빼면 된다.
remains = result[0] - 1760

print('남은 돈은 {}원 입니다.'.format(remains))
프로그램을 실행한 결과는 다음과 같다.

남은 돈은 2640원입니다.

#### sympy 활용
x^2 = 1과 같은 이차방정식의 해를 구해보자.

In [31]:
import sympy
x = sympy.symbols('x')
f = sympy.Eq(x**2, 1)
sympy.solve(f)

[-1, 1]

또한, 다음과 같은 연립방정식의 해도 구할 수 있다.  
x + y = 10  
x - y = 4

In [33]:
x, y = sympy.symbols('x y')
f1 = sympy.Eq(x+y, 10)
f2 = sympy.Eq(x-y, 4)
sympy.solve([f1, f2])

{x: 7, y: 3}

미지수가 2개 이상이라면 결괏값이 리스트가 아닌 딕셔너리임에 주의하자.